### PR TITLE
Implement functionality to create a topic and pubsub client, and then publish a message to the given topic

### PR DESCRIPTION
**Pull request by DavidHODs**

### CODE REVIEW

**Comment by DavidHODs** [comment]
"When you create the function that creates this topic, it should return a callback function which the user can call to trigger traffic.Stop()

Why all this is important, is that we will be publishing messages hundreds of thousands of times per minute. And we don't want to be initializing a topic for each message. We want to initialize the topic once, and then reuse it every time we need to publish the message. Then when everything is completely done, the caller can shutdown the topic"

I missed this in your previous reviews..I was doing a final round check before making the latest push when I saw this. I'll push an updated workflow by end of today or midnight to cater for this

**Comment by tonyalaribe** [comment]
Maybe next time, make the changes within the same branch you're working on. Closing and recreating the pullrequest is not a good approach since you're basically throwing away the history

File: sdk.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+func objectJSON(payload string) (map[string]interface{}, error) {
+	var result map[string]interface{}
+
+	err := json.Unmarshal([]byte(payload), &result)
+	if err != nil {
+		return nil, err
+	}
+
+	return result, err
+}
```
Comment (tonyalaribe) [inline sdk.go:67]: Please don't create a separate function for this.  Just marshal the code where it's needed.
Also, you should be marshalling, not unmarshalling.

File: sdk.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+// PublishMessage publishes payload to a gcp cloud console 
+func PublishMessage(ctx context.Context, payload string) error {
```
Comment (tonyalaribe) [inline sdk.go:70]: You had an object before that represented the payload. Why did you delete it?
Please bring it back. The payload is an object that represents the requests and response and  not a string.

File: sdk.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+	topic, err := initializeTopic(ctx)
+	if err != nil {
+		return err
+	}
+
+	data, err := objectJSON(payload)
+	if err != nil {
+		return err
+	}
```
Comment (tonyalaribe) [inline sdk.go:79]: Marshal the payload in this function. It's 2 lines of code. No need to move that into another function.

File: sdk.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+	msgg := &pubsub.Message{
+		ID:              ProjectID,
+		Data:            []byte(fmt.Sprintf("payload: %v", data)),
```
Comment (tonyalaribe) [inline sdk.go:83]: When you marshal the object, you should use the byte data here directly. Not wrapping it as test
```suggestion
		Data:            data,
```

File: sdk.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+		PublishTime:     time.Now(),
+	}
+
+	topic.Publish(ctx, msgg).Get(ctx)
+	defer topic.Stop()
```
Comment (tonyalaribe) [inline sdk.go:88]: Don't Stop() the topic after publishing. By doing this, you make it such that the topic cannot be reused.
Like I said in the last PR, the topic should be reused because we will be publishing thousands of times per minute. 
So the user of the library should be the one stopping the topic when they are done or when their app is shutting down. You just need to provide the logic to make that possible

File: sdk_test.go [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+	if err != nil {
+		t.Error(err)
+	}
+
+	if !exists {
+		_, err = client.CreateTopic(context.Background(), TopicID)
+		if err != nil {
+			t.Error(err)
+		}
+	}
```
Comment (tonyalaribe) [inline sdk_test.go:39]: Shouldn't you be testing your functions? Why are you testing the pubsub library  functions  instead of your own functions?

File: sdk_test.go [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+	"cloud.google.com/go/pubsub"
+	"github.com/joho/godotenv"
+)
+
+func TestInitializeClient(t *testing.T) {
+	_ = godotenv.Load(".env")
+	_, err := pubsub.NewClient(context.Background(), ProjectID)
+	if err != nil {
+		t.Error(err)
+	}
```
Comment (tonyalaribe) [inline sdk_test.go:19]: Again, why are you testing the pubsub library instead of the functions you wrote?
I genuinely don't understand what's going on. Please explain

File: sdk_test.go [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+	err := json.Unmarshal([]byte(`{
+		"title": "blog post",
+		"body": "this is a blog post",
+		"likes": "98",
+		"comments": "45"
+	}`), &result)
+
+	if err != nil {
+		t.Error(t)
+	}
```
Comment (tonyalaribe) [inline sdk_test.go:54]: Again, what are you testing here?

**Review by tonyalaribe — CHANGES_REQUESTED** [review]

File: sdk_test.go (hunk 7 above)
Comment (DavidHODs) [reply sdk_test.go:19]: I'll redo the testing

No explanation can justify what I did.

I'm sorry for the below average work

Comment (tonyalaribe) [reply sdk_test.go:19]: Man, it’s not that deep. Hahaha. 
just call the functions in the test, instead of copying the implementation. 

Also, use an assertion library to assert values in your tests: https://github.com/stretchr/testify

File: sdk.go [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+
+// PublishMessage publishes payload to a gcp cloud console 
+func PublishMessage(ctx context.Context, payload Data) (error) {
+	topic, err := initializeTopic(ctx)
+	if err != nil {
+		return err
+	}
+
```
Comment (tonyalaribe) [inline sdk.go:71]: Can you pass in the topic as an argument. So we don't initialize topic everytime we create a message

File: sdk.go [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+	topic, err := client.CreateTopic(ctx, TopicID)
+		if err != nil {
+			return nil, err
+		}
+
+	return topic, err
+}
+
+// initializes the topic instance
+var topicInstance, errTopicInstance = initializeTopic(context.Background())
```
Comment (tonyalaribe) [inline sdk.go:65]: Accept the topic as an argument. Relying on global variables is bad practice and is an engineering code smell
